hedera-node/src/main/java/com/hedera/services/txns/crypto/CryptoApproveAllowanceTransitionLogic.java

```diff
@@ -0,0 +1,213 @@
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.MAX_ALLOWANCES_EXCEEDED;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.SUCCESS;
+
+public class CryptoApproveAllowanceTransitionLogic implements TransitionLogic {
+	private final TransactionContext txnCtx;
+	private final SigImpactHistorian sigImpactHistorian;
+	private final AccountStore accountStore;
+	private final AllowanceChecks allowanceChecks;
+
+	// Should these be dynamic properties?
+	public static final int ALLOWANCE_LIMIT_PER_TRANSACTION = 20;
+	public static final int TOTAL_ALLOWANCE_LIMIT_PER_ACCOUNT = 100;
+
+	@Inject
+	public CryptoApproveAllowanceTransitionLogic(
+			final TransactionContext txnCtx,
+			final SigImpactHistorian sigImpactHistorian,
+			final AccountStore accountStore,
+			final AllowanceChecks allowanceChecks) {
+		this.txnCtx = txnCtx;
+		this.sigImpactHistorian = sigImpactHistorian;
+		this.accountStore = accountStore;
+		this.allowanceChecks = allowanceChecks;
+	}
+
+	@Override
+	public void doStateTransition() {
+		/* --- Extract gRPC --- */
+		final TransactionBody cryptoApproveAllowanceTxn = txnCtx.accessor().getTxn();
+		final AccountID owner = cryptoApproveAllowanceTxn.getTransactionID().getAccountID();
+		final var op = cryptoApproveAllowanceTxn.getCryptoApproveAllowance();
+
+		/* --- Use models --- */
+		final Id ownerId = Id.fromGrpcAccount(owner);
+		final var ownerAccount = accountStore.loadAccount(ownerId);
+
+		/* --- Do the business logic --- */
+		applyCryptoAllowances(op.getCryptoAllowancesList(), ownerAccount);
+		applyFungibleTokenAllowances(op.getTokenAllowancesList(), ownerAccount);
+		applyNftAllowances(op.getNftAllowancesList(), ownerAccount);
+
+		/* --- validate --- */
+		validateFalse(exceedsAccountLimit(ownerAccount), MAX_ALLOWANCES_EXCEEDED);
+
+		/* --- Persist the owner account --- */
+		accountStore.commitAccount(ownerAccount);
+		sigImpactHistorian.markEntityChanged(ownerId.num());
+
+		txnCtx.setStatus(SUCCESS);
+	}
+
+	@Override
+	public Predicate<TransactionBody> applicability() {
+		return TransactionBody::hasCryptoApproveAllowance;
+	}
+
+	@Override
+	public Function<TransactionBody, ResponseCodeEnum> semanticCheck() {
+		return this::validate;
+	}
+
+	private ResponseCodeEnum validate(TransactionBody cryptoAllowanceTxn) {
+		final AccountID owner = cryptoAllowanceTxn.getTransactionID().getAccountID();
+		final var ownerAccount = accountStore.loadAccount(Id.fromGrpcAccount(owner));
+		return allowanceChecks.allowancesValidation(cryptoAllowanceTxn, ownerAccount);
+	}
+
+	/**
+	 * Applies all changes needed for Crypto allowances from the transaction
+	 *
+	 * @param cryptoAllowances
+	 * @param ownerAccount
+	 */
+	private void applyCryptoAllowances(final List<CryptoAllowance> cryptoAllowances, final Account ownerAccount) {
+		if (cryptoAllowances.isEmpty()) {
+			return;
+		}
+		Map<EntityNum, Long> cryptoAllowancesMap = ownerAccount.getCryptoAllowances();
```

Comment:
E.g. here I might have expected to see instead,
```
Map<EntityNum, Long> cryptoAllowancesMap = new TreeMap<>(ownerAccount.getCryptoAllowances());
```
so that we are creating the new `TreeMap` only when it's needed.

Reply:
Makes sense. Made the change